firmware/controllers/engine_cycle/high_pressure_fuel_pump.cpp

```diff
@@ -13,70 +13,172 @@
 #include "high_pressure_fuel_pump.h"
 #include "spark_logic.h"
 
-#if EFI_HPFP
-
-#define LOBE_COUNT 3
+//#if EFI_HPFP
+#if 1
 
 class HpfpActor {
 public:
-	angle_t extra;
-	int phaseIndex;
+    uint8_t m_lobe_index;
+    float m_I_sum;
+    float m_pressureTarget;
+
+    AngleBasedEvent m_open;
+    AngleBasedEvent m_close;
 
-	AngleBasedEvent open;
-	AngleBasedEvent close;
 	DECLARE_ENGINE_PTR;
+
+    void start();
+
+private:
+    void scheduleNextCycle();
+    angle_t findNextLobe();
+    float calcFuelPercent(int rpm);
+    float calcPI(int rpm, float calc_fuel_percent);
+
+    static void hpfpPinTurnOn(HpfpActor *) {
+        enginePins.hpfpValve.setHigh();
+    }
+
+    static void hpfpPinTurnOff(HpfpActor *current) {
+        RegisteredNamedOutputPin *output = &enginePins.hpfpValve;
+        output->setLow();
+
+        current->scheduleNextCycle();
+    }
 };
 
-static HpfpActor actors[LOBE_COUNT];
+static HpfpActor actor; // TODO: Instantiate one per bank? Need more sensors....
+
+template<int TColNum, int TRowNum, typename vType, typename kType,
+         typename TValueMultiplier = efi::ratio<1>>
+auto makeMap3D(vType table[TRowNum][TColNum], kType rowBins[TRowNum], kType columnBins[TColNum]) {
```

Comment:
this will be unnecessary after https://github.com/rusefi/rusefi/pull/3452

Reply:
How so?  It isn't immediately obvious to me.  FYI the point of this change is to just create the Map3D structure on the fly, rather than precreating it.  Precreating has no performance advantage, it takes 4 words of RAM, and it actually slower due to the vtable indirection (though that could be changed by marking some of the overridden methods as final).